synapse/handlers/device.py

```diff
@@ -1300,6 +1304,12 @@ async def _user_device_resync_returning_failed(
               e.g. due to a connection problem.
             - True iff the resync failed and the device list should be marked as stale.
         """
+        # Check that we haven't gone and fetched the devices since we last
+        # checked if we needed to resync these device lists.
+        if await self.store.get_users_whose_devices_are_cached([user_id]):
+            cached = await self.store.get_cached_devices_for_user(user_id)
+            return cached, False
```

Comment:
This is safe because if we linearized on user ID, then request the same user again but have *not* received a request to invalidate that data (via a resync request) then it must still be valid?

Reply:
Oh, we already do do this check *before* entering this block, so this is safe. The flow is:

1. Client reader receives a requests for keys.
2. Client reader checks if keys are cached, if so returns them.
3. If not, client reader queries master to fetch the keys.
4. **New** Master linearizes requests for a users keys
5. **New** Master checks if we *now* have the keys in the cache, if so returns.
6. Fetches keys from the remote server, and caches them if we share a room with the remote user.

So this is step 5, which is safe because we're basically re-doing a check we have previously done